core/signeddata.go

```diff
@@ -0,0 +1,277 @@
+		resp.Phase0.Signature = sig.ToETH2()
+	case spec.DataVersionAltair:
+		resp.Altair.Signature = sig.ToETH2()
+	case spec.DataVersionBellatrix:
+		resp.Bellatrix.Signature = sig.ToETH2()
+	default:
+		return nil, errors.New("unknown type")
+	}
+
+	return resp, nil
+}
+
+func (b VersionedSignedBeaconBlock) MarshalJSON() ([]byte, error) {
+	var marshaller json.Marshaler
+	switch b.Version {
+	// No block nil checks since `NewVersionedSignedBeaconBlock` assumed.
+	case spec.DataVersionPhase0:
+		marshaller = b.VersionedSignedBeaconBlock.Phase0
+	case spec.DataVersionAltair:
+		marshaller = b.VersionedSignedBeaconBlock.Altair
+	case spec.DataVersionBellatrix:
+		marshaller = b.VersionedSignedBeaconBlock.Bellatrix
+	default:
+		return nil, errors.New("unknown version")
+	}
+
+	block, err := marshaller.MarshalJSON()
+	if err != nil {
+		return nil, errors.Wrap(err, "marshal block")
+	}
+
+	resp, err := json.Marshal(versionedSignedBeaconBlockJSON{
+		Version: int(b.Version),
+		Block:   block,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "marshal wrapper")
+	}
+
+	return resp, nil
+}
+
+func (b *VersionedSignedBeaconBlock) UnmarshalJSON(input []byte) error {
+	var raw versionedSignedBeaconBlockJSON
+	if err := json.Unmarshal(input, &raw); err != nil {
+		return errors.Wrap(err, "unmarshal block")
+	}
+
+	resp := spec.VersionedSignedBeaconBlock{Version: spec.DataVersion(raw.Version)}
+	switch resp.Version {
+	case spec.DataVersionPhase0:
+		block := new(eth2p0.SignedBeaconBlock)
+		if err := json.Unmarshal(raw.Block, &block); err != nil {
+			return errors.Wrap(err, "unmarshal phase0")
+		}
+		resp.Phase0 = block
+	case spec.DataVersionAltair:
+		block := new(altair.SignedBeaconBlock)
+		if err := json.Unmarshal(raw.Block, &block); err != nil {
+			return errors.Wrap(err, "unmarshal altair")
+		}
+		resp.Altair = block
+	case spec.DataVersionBellatrix:
+		block := new(bellatrix.SignedBeaconBlock)
+		if err := json.Unmarshal(raw.Block, &block); err != nil {
+			return errors.Wrap(err, "unmarshal bellatrix")
+		}
+		resp.Bellatrix = block
+	default:
+		return errors.New("unknown version")
+	}
+
+	b.VersionedSignedBeaconBlock = resp
+
+	return nil
+}
+
+// versionedSignedBeaconBlockJSON is a custom VersionedSignedBeaconBlock serialiser.
+type versionedSignedBeaconBlockJSON struct {
+	Version int             `json:"version"`
+	Block   json.RawMessage `json:"block"`
+}
+
+// NewAttestation is a convenience function that returns a new wrapped attestation.
+func NewAttestation(att *eth2p0.Attestation) Attestation {
+	return Attestation{Attestation: *att}
+}
+
+// NewPartialAttestation is a convenience function that returns a new partially signed attestation.
+func NewPartialAttestation(att *eth2p0.Attestation, shareIdx int) ParSignedData2 {
+	return ParSignedData2{
+		SignedData: NewAttestation(att),
+		ShareIdx:   shareIdx,
+	}
+}
+
+// Attestation is a signed attestation and implements SignedData.
+type Attestation struct {
+	eth2p0.Attestation
```

Comment:
cannot subtype since `eth2p0.Attestation`'s field `Signature` clashes with interface method `Signature`

Reply:
can we make this attestation private since we already have getter and setter methods(Signature and SetSignature)?

Reply:
Although I don't have any strong opinion on this.

Reply:
it cannot be made private since some components will cast and use it, like bcast, and aggsigdb